urllib3/connectionpool.py

```diff
@@ -687,12 +687,17 @@ def urlopen(self, method, url, body=None, headers=None, retries=None,
                 retries = retries.increment(method, url, response=response, _pool=self)
             except MaxRetryError:
                 if retries.raise_on_redirect:
-                    # Release the connection for this response, since we're not
-                    # returning it to be released manually.
+                    # Drain and release the connection for this response, since
+                    # we're not returning it to be released manually.
+                    response.read()
                     response.release_conn()
```

Comment:
@mattbillenstein Yeah, so I had that thought too. I think it's possible that `release_conn` should check connection state. However, I think for `release_conn` the calculation is different: if the response is not complete when `release_conn` is called, probably the connection should be *closed*, not drained. Draining the connection here is still a specific-to-this-use-case operation, because here we have confidence that the connection is fine. However, when a user calls `release_conn` for any reason, we have no idea what's going on, and don't have any good way to keep track of what we think is going on.

So, I'd consider the enhancement to `release_conn` to be a separate concern, IMO: in this case we'll still want to drain the connection.